docs: improve intro to ops.testing reference

dwilding opened the pull request:
Resolves #1826. This PR makes a few small improvements to the intro of the `ops.testing` reference:

- Mention that how-to guides have info about testing each feature
- Link to the unit testing how-to guide and the testing explanation (moved from the end of the intro)
- Mention the term "unit tests" much earlier
- Move the "previously known as 'Scenario'" part to the end of the intro, as Scenario is already mentioned in the title

**[Preview doc](https://canonical-ubuntu-documentation-library--2023.com.readthedocs.build/ops/2023/reference/ops-testing/)**

---

Not implemented:

- In [#1826](https://github.com/canonical/operator/issues/1826), I mentioned possibly adding a link from [`Relation`](https://documentation.ubuntu.com/ops/latest/reference/ops-testing/#ops.testing.Relation) to the how-to guide. But why only from `Relation`? I'm adding the new info right at the beginning of the intro, so let's first see how well that works.

- I also mentioned using full URLs for links, so that the links work in docstrings. We don't actually need to do that here. The docstring of `ops.testing` comes from `ops/testing.py` not the intro of `docs/reference/ops-testing.rst` - that intro is purely for the docs.

james-garner-canonical commented on docs/reference/ops-testing.rst:
```diff
@@ -6,7 +6,15 @@
+- :doc:`/howto/write-unit-tests-for-a-charm`
+- :doc:`/explanation/testing` - A summary of types of charm tests
+
+State-transition tests, also called 'unit' tests, expect you to define the
```
This reads as defining 'unit tests' === 'state-transition tests'. The intent of the previous version of this line was to say 'scenario' === 'state-transition tests'. If the intent is to remove references to 'scenario', how about 'state-transition unit tests'? Otherwise perhaps mention `ops-scenario`/`ops[testing]`/etc?

dwilding replied:
Good point, the terms are getting a bit confused here. How about if I modify the first paragraph about state-transition tests and include a second paragraph:

> State-transition tests expect you to define the Juju state all at once, define the Juju context against which to test the charm, and fire a single event on the charm to execute its logic. The tests can then assert that the Juju state has changed as expected.
>
> Unlike integration tests, state-transition tests do not test using a real Juju controller and model. They focus on a single Juju unit. For this reason, state-transition tests are a form of 'unit' tests. In fact, they are the primary form of unit tests that you should write for a charm.

For the second paragraph, I'm moving & adapting some wording from earlier in the intro.

dwilding replied:
Nice, I like this, thanks. Updated: https://canonical-ubuntu-documentation-library--2023.com.readthedocs.build/ops/2023/reference/ops-testing/

IronCore864 commented:
It would be nice if the "Test the feature" part worked as a link, but I played with it a little bit, and it doesn't...

dwilding commented:
> It would be nice if the "Test the feature" part worked as a link

Good idea, I've added a cross-reference target so that it can become a link. Updated: https://canonical-ubuntu-documentation-library--2023.com.readthedocs.build/ops/2023/reference/ops-testing/
